### PR TITLE
feat: Add Query Performance Monitoring (QPM) support for MariaDB

### PR DESCRIPTION
**Pull request by RamanaReddy8801**
## Overview
This PR adds comprehensive Query Performance Monitoring (QPM) support for MariaDB databases (version 10.2 and above), enabling the MySQL integration to collect and report query performance metrics from MariaDB instances. Additionally, implements enhanced deduplication logic to prevent duplicate query metrics across overlapping Performance Schema tables.

### Database Flavor Detection
- Added `DatabaseFlavor` type with `DatabaseFlavorMySQL` and `DatabaseFlavorMariaDB` constants
- Implemented `DetectDatabaseFlavor()` function to automatically identify the database type from version strings
- Enhanced `DatabaseProfile` struct to store both flavor and raw version information

### Version Validation
- Added MariaDB version validation with minimum support for MariaDB 10.2+ (required for CTE/WITH clause support)
- Implemented `isMariaDBVersionSupported()` function with proper version parsing
- Added version constants: `MinMariaDBMajorVersion = 10`, `MinMariaDBMinorVersion = 2`
- Graceful handling of version suffixes (e.g., "10.6.12-MariaDB", "10.2-MariaDB")

### MariaDB-Specific Query Adaptations
- **Slow Queries**: Created `MariaDBSlowQueries` variant that sets CPU time to NULL (field not consistently available across MariaDB versions)
- **Blocking Sessions**: Added query normalization for MariaDB since it lacks pre-normalized DIGEST_TEXT
- **Query Sets**: Implemented dynamic query set selection based on database flavor via `GetQuerySet()` function

### Query Execution Plan Handling
Enhanced execution plan metrics extraction to handle MariaDB-specific JSON format differences:
- **Duplicate Key Detection**: Added `deduplicateJSONKeys()` to handle MariaDB 10.6 JOIN output with duplicate "table" keys
- **Flexible Field Parsing**: Supports multiple field formats across MariaDB versions:
  - `filtered` field as both string and number types
  - `rows` field with fallback to `r_rows` (MariaDB specific)
  - Flat `cost` field in MariaDB 11.4+
- Pre-processes raw JSON before unmarshaling to ensure compatibility

### Helper Functions
- Added `NormalizeQueryText()` function to normalize MariaDB query texts (stripping literals, collapsing whitespace)
- Enhanced query preprocessing and validation logic

### Enhanced Duplicate Prevention
**Implemented robust deduplication for individual query metrics:**
  - Added `deduplicateIndividualQueryMetrics()` function using EVENT_ID + THREAD_ID + EXECUTION_TIME combination
  - Prevents duplicate query executions from being reported when they appear across multiple Performance Schema tables (events_statements_current, events_statements_history, events_statements_history_long)
  - Supports multi-application scenarios with shared or separate connection pools
  - Uses millisecond precision (%.3f) for execution time comparison to distinguish between genuinely different executions

### Deduplication Test Coverage
  **Comprehensive test suite for enhanced deduplication logic:**
  - Empty input and nil value handling (EVENT_ID, THREAD_ID, ExecutionTimeMs)
  - True duplicate detection (identical EVENT_ID + THREAD_ID + execution time)
  - Edge case scenarios (same IDs with different execution times)
  - Multi-application scenarios:
    - Separate connection pools (different THREAD_IDs)
    - Shared connection pools (same THREAD_ID, different EVENT_IDs)
    - Connection pool recycling with thread reuse

### Testing
Added comprehensive test coverage with 18 new test cases:
- MariaDB version validation tests (including edge cases)
- Query execution plan tests for MariaDB 10.6, 11.4+
- Blocking sessions tests with query normalization
- Query text normalization tests
- Version string parsing tests

### Why MariaDB 10.2 Minimum?
The binding constraint is the `WaitEventsQuery` which uses CTEs (WITH clause) introduced in MariaDB 10.2. Other collectors support lower versions:
- Slow queries: 10.0 (events_statements_* tables)
- Execution plans: 10.1 (EXPLAIN FORMAT=JSON)
- Blocking sessions: 10.0 (information_schema.innodb_lock_waits)
- **Wait events: 10.2** (CTE/WITH clause) ← minimum

## Files Modified
- Core QPM files: validators, query collectors, performance main
- Utilities: database profile, helpers, queries, query sets
- Performance collectors: blocking sessions, query details, query execution plan
- Constants: added MariaDB version requirements
- Tests: added extensive MariaDB-specific test coverage

### Test Infrastructure

- Docker Compose: Added 3 MariaDB containers to docker-compose-performance.yml
     - mariadb_perf_10-6 (MariaDB 10.6.23)
    - mariadb_perf_11-4 (MariaDB 11.4.10)
    - mariadb_perf_12-2 (MariaDB 12.2.2)

- **Dockerfiles**: Created version-specific Dockerfiles with performance schema enabled and test database setup
## Test Coverage

- **Version Matrix**: All QPM tests now run against both MySQL (8.0.40, 8.4.0, 9.1.0) and MariaDB (10.6, 11.4, 12.2) versions
- **Automated Flavor Detection**: Tests automatically select appropriate JSON schemas based on detected database flavor
- Full QPM Metrics Validation:

    - Slow query metrics
    - Individual query performance metrics
    - Query execution plans (including MariaDB-specific JSON format handling)
    - Wait event metrics
    - Blocking sessions (with query normalization for MariaDB)

### Created 4 MariaDB-specific JSON schema files for validation:

- [mariadb-schema-master.json] - Remote entity with QPM metrics
- [mariadb-schema-master-localentity.json] - Local entity with QPM metrics
- [mariadb-schema-inventory-master.json] - Inventory data validation
- [mariadb-schema-slow-queries.json] - Slow query schema (CPU time set to NULL for MariaDB compatibility)
### Implemented conditional schema selection in test suite:

- MariaDB tests use MariaDB-specific schemas for the 4 files with differences
- All other metrics (individual queries, execution plans, wait events, blocking sessions) use shared MySQL schemas since formats are identical
- Graceful handling of missing/empty output for version-specific features

## Co-authored-by
Co-authored-by: pkudikyala <pkudikyala@newrelic.com>
Co-authored-by: Rahul Malhan <rmalhan@newrelic.com>

### CODE REVIEW

File: src/query-performance-monitoring/utils/helpers.go [hunk 1]
```diff
@@ -24,6 +25,28 @@ var (
...
+// literalNormalizer matches SQL literal values in a single pass (same approach as nri-mssql
+// AnonymizeQueryText). Priority order matters: strings are matched first so numbers inside
+// quoted values are not double-replaced. Handles:
+//  1. Single-quoted strings — both SQL ('it''s') and MySQL backslash ('O\'Brien') escape styles
+//  2. Hex literals — 0xDEADBEEF
+//  3. Numeric literals — integers and decimals, word-bounded to skip numbers in identifiers (col_1)
+//
+// Used only for MariaDB blocking queries; MySQL DIGEST_TEXT is already normalized by
+// performance_schema so no Go-side normalization is needed there.
+var literalNormalizer = regexp.MustCompile(`'(?:[^'\\]|\\.|'')*'|0x[0-9a-fA-F]+|\b\d+(?:\.\d+)?\b`)
```
Comment (spathlavath) [inline src/query-performance-monitoring/utils/helpers.go]: Rename `literalNormalizer` to `literalAnonymizer`. Also Update the comment accordingly. can you please update this in all the places.

File: src/query-performance-monitoring/utils/helpers.go [hunk 2]
```diff
@@ -24,6 +25,28 @@ var (
...
+
+// NormalizeQueryText replaces string literals, hex values, and numeric literals
+// in a raw SQL query with ? placeholders, matching performance_schema DIGEST_TEXT format.
+// Returns nil if the input is nil.
+func NormalizeQueryText(query *string) *string {
```
Comment (spathlavath) [inline src/query-performance-monitoring/utils/helpers.go]: same as above.

File: src/query-performance-monitoring/utils/query_sets.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// MariaDB uses modified queries for slow queries and blocking sessions
+func GetQuerySet(flavor DatabaseFlavor) QuerySet {
+	slowQuery := SlowQueries
+	blockingQuery := BlockingSessionsQuery
+	needsNormalization := false
+
+	if flavor == DatabaseFlavorMariaDB {
+		slowQuery = MariaDBSlowQueries
+		blockingQuery = MariaDBBlockingSessionsQuery
+		needsNormalization = true
```
Comment (spathlavath) [inline src/query-performance-monitoring/utils/query_sets.go]: same here. it's Anonymization.

**Comment by spathlavath** [comment]
Please add integration tests.

File: src/query-performance-monitoring/performance-metrics-collectors/query_details.go [hunk 4]
```diff
@@ -146,7 +146,44 @@ func getIndividualQueryList(db utils.DataSource, queryIDList []string, args argu
...
+
+	for _, metric := range metrics {
+		// Skip metrics with nil fields - these shouldn't occur due to query structure
+		if metric.EventID == nil || metric.ThreadID == nil || metric.ExecutionTimeMs == nil {
+			continue
+		}
+
+		// Always use enhanced key with execution time for robust deduplication
+		// This handles edge cases where EVENT_ID + THREAD_ID might be reused
+		key := fmt.Sprintf("%d_%d_%.3f", *metric.EventID, *metric.ThreadID, *metric.ExecutionTimeMs)
```
Comment (spathlavath) [inline src/query-performance-monitoring/performance-metrics-collectors/query_details.go]: The value is already rounded to 3 decimal places at the SQL level, so can please remove this `%.3f`

Comment (spathlavath) [inline src/query-performance-monitoring/performance-metrics-collectors/query_details.go]: `EVENT_ID` is an auto-incrementing counter per thread in Performance Schema - it is not reused within a monitoring window. The same (`EVENT_ID, THREAD_ID`) pair always identifies exactly `one execution`. Adding `ExecutionTimeMs` to the key doesn't protect against reuse; it instead allows genuine duplicates to slip through if there's any float representation difference between the three table reads.
> key := fmt.Sprintf("%d_%d", *metric.EventID, *metric.ThreadID)

File: src/query-performance-monitoring/performance-metrics-collectors/query_details.go [hunk 5]
```diff
@@ -146,7 +146,44 @@ func getIndividualQueryList(db utils.DataSource, queryIDList []string, args argu
...
+		return metrics
+	}
+
+	// Use a map to track unique combinations
+	seen := make(map[string]bool)
+	var deduplicated []utils.IndividualQueryMetrics
+
+	for _, metric := range metrics {
+		// Skip metrics with nil fields - these shouldn't occur due to query structure
+		if metric.EventID == nil || metric.ThreadID == nil || metric.ExecutionTimeMs == nil {
```
Comment (spathlavath) [inline src/query-performance-monitoring/performance-metrics-collectors/query_details.go]: can you please add warning log here, so that it will be helpful while debugging.
> log.Warn("Skipping individual query metric with nil EventID/ThreadID/ExecutionTimeMs - possible scan error")

File: src/query-performance-monitoring/utils/queries.go
Comment (spathlavath) [inline src/query-performance-monitoring/utils/queries.go]: @RamanaReddy8801 `PastQueriesSearch (events_statements_history_long)` is a superset of `RecentQueriesSearch (events_statements_history)` — both tables contain only ended events, and every row in history also exists in history_long. Querying both is the primary source of duplicates between these two tables. Consider removing `PastQueriesSearch` entirely: history (last 10 per thread) is sufficient for fetching execution plan samples for recently slow queries, and it eliminates the most expensive of the three queries (up to 10,000 rows globally on a busy server).

File: src/query-performance-monitoring/validator/validations.go [hunk 6]
```diff
@@ -223,12 +232,12 @@ func checkAndEnableEssentialConsumers(db utils.DataSource) error {
...
 
 // logEnablePerformanceSchemaInstructions logs instructions to enable the Performance Schema.
-func logEnablePerformanceSchemaInstructions(version string) {
-	if isVersion8OrGreater(version) {
-		log.Debug("To enable the Performance Schema, add the following lines to your MySQL configuration file (my.cnf or my.ini) in the [mysqld] section and restart the MySQL server:")
+func logEnablePerformanceSchemaInstructions(profile utils.DatabaseProfile) {
+	if profile.Flavor == utils.DatabaseFlavorMariaDB || isVersion8OrGreater(profile.RawVersion) {
+		log.Debug("To enable the Performance Schema, add the following lines to your database configuration file (my.cnf or my.ini) in the [mysqld] section and restart the server:")
 		log.Debug("performance_schema=ON")
 	} else {
```
Comment (spathlavath) [inline src/query-performance-monitoring/validator/validations.go]: @RamanaReddy8801 The else branch is dead code - by the time this function is called, unsupported MySQL versions are already rejected in ValidatePreconditions. The `log.Error` here can never be reached. Consider removing the else branch entirely.

**Review by spathlavath — APPROVED** [review]
LGTM, Please make sure it shouldn't break any existing functionality.
